lang_java/tree_sitter/codegen.py

```diff
@@ -0,0 +1,109 @@
+import json
+
+with open("node-types.json", 'r') as fs:
+    data = json.load(fs)
+
+codegen = '''
+(* 
+ * Yoann Padioleau, Sharon Lin
+ * 2020 initial draft
+ *)
+
+(*****************************************************************************)
+(* Prelude *)
+(*****************************************************************************)
+(*
+ * An AST for Java.
+ *
+ * todo:
+ *  - finish ast import from 
+ *    https://github.com/tree-sitter/tree-sitter-java/blob/master/grammar.js
+ *)
+ (*****************************************************************************)
+(* The Tree-sitter AST java related types *)
+(*****************************************************************************)
+(* ------------------------------------------------------------------------- *)
+(* Token/info *)
+(* ------------------------------------------------------------------------- *)
+
+type tok = Parse_info.t
+  (* with tarzan *)
+type 'a wrap  = 'a * tok
+  (* with tarzan *)
+
+type 'a list1 = 'a list (* really should be 'a * 'a list *)
+  (* with tarzan *)
+'''
+
+for a in data:
+    if 'subtypes' in a:
+        codegen += '''
+type {type} = '''.format(type=a['type'])
+        for b in a['subtypes']:
+            if len(b['type'].split("_")[0]) == 0:
+                constr = b['type'].split("_")[1]
+            else:
+                constr = b['type'].split("_")[0]
+
+            codegen+='''
+| {constructor} of {field}'''.format(constructor=constr.capitalize(),field=b['type'])
+        codegen += "\n"
+    
+    
+    elif 'fields' in a and len(a['fields']) > 0:
+        codegen += '''
+and {type} = {{'''.format(type=a['type'])
+        for b, bval in a['fields'].items():
+            codegen += '''
+  {field}: '''.format(field=b)
+            for c in bval['types']:
+                codegen +="{param} ".format(param=c['type'])
+                if bval['multiple'] == True:
+                    codegen += "list "  
+                if bval['required'] == False:
+                    codegen += "option "  
+                codegen += "* "
+            codegen = codegen[:-3] + ";"
+        
+        if 'children' in a:
+            codegen +='''
+  mods: '''
+            for b in a['children']['types']:
+                codegen +="{mod} ".format(mod=b['type'])
+                if a['children']['multiple'] == True:
+                    codegen += "list "
+                if a['children']['required'] == False:
+                    codegen += "option "
+                codegen += "| "
+            codegen = codegen[:-3] + ";\n"
+        codegen += "\n}\n"
+    
+    
+    elif 'children' in a:
+        codegen+='''
+and {param} = '''.format(param=a['type'])
+        for b in a['children']['types']:
+            if len(b['type'].split("_")[0]) == 0:
```

Comment:
Ditto about the startswith check. Honestly, if you use the same code twice, better use a helper function for it